File: backend/app/api/v1/memories.py
"""MemoryHub routes."""

from __future__ import annotations

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.core.deps import DbSession, get_current_user
from app.models.memory import Memory
from app.models.user import User
from app.schemas.memory import (
    MemoryKind,
    MemoryList,
    MemoryOut,
    MemoryScopeType,
    MemoryStatus,
    UpdateMemoryRequest,
)
from app.services.memory_hub import MemoryHubService

router = APIRouter()


@router.get("", response_model=MemoryList)
async def list_memories(
    db: DbSession,
    user: Annotated[User, Depends(get_current_user)],
    scope_type: MemoryScopeType | None = Query(default=None),
    scope_id: UUID | None = Query(default=None),
    kind: MemoryKind | None = Query(default=None),
    status_filter: MemoryStatus | None = Query(default=None, alias="status"),
    limit: int = Query(default=50, ge=1, le=200),
) -> MemoryList:
    items, total = await MemoryHubService().list_memories(
        db,
        owner_user_id=user.id,
        scope_type=scope_type,
        scope_id=scope_id,
        kind=kind,
        status=status_filter,
        limit=limit,
    )
    return MemoryList(
        items=[MemoryOut.model_validate(item) for item in items],
        total=total,
    )


@router.patch("/{memory_id}", response_model=MemoryOut)
async def update_memory(
    memory_id: UUID,
    payload: UpdateMemoryRequest,
    db: DbSession,
    user: Annotated[User, Depends(get_current_user)],
) -> MemoryOut:
    memory = await _get_owned_memory(db, user.id, memory_id)
    updated = await MemoryHubService().update_memory(
        db,
        memory,
        content=payload.content,
        importance=payload.importance,
        status=payload.status,
        kind=payload.kind,
        valid_until=payload.valid_until,
        metadata=payload.metadata,
    )
    await db.commit()
    await db.refresh(updated)
    return MemoryOut.model_validate(updated)


@router.delete("/{memory_id}", response_model=MemoryOut)
async def forget_memory(
    memory_id: UUID,
    db: DbSession,
    user: Annotated[User, Depends(get_current_user)],
) -> MemoryOut:
    memory = await _get_owned_memory(db, user.id, memory_id)
    updated = await MemoryHubService().forget_memory(db, memory)
    await db.commit()
    await db.refresh(updated)
    return MemoryOut.model_validate(updated)


async def _get_owned_memory(db: DbSession, user_id: UUID, memory_id: UUID) -> Memory:
    memory = await db.get(Memory, memory_id)
    if memory is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"error": {"code": "MEMORY_NOT_FOUND", "message": "Not found"}},
        )
    if memory.owner_user_id != user_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={"error": {"code": "FORBIDDEN", "message": "Forbidden"}},
        )
    return memory

File: frontend/src/lib/adapters/memories.ts
import { api } from '@/lib/api';
import type {
  Memory,
  ConversationMemoryHub,
  MemoryList,
  MemoryMountList,
  UpdateMemoryRequest,
} from '@/lib/types';

export type MemoryScopeType = Memory['scope_type'];
export type MemoryKind = Memory['kind'];
export type MemoryStatus = Memory['status'];

export async function listMemories(params: {
  scopeType?: MemoryScopeType;
  scopeId?: string;
  kind?: MemoryKind;
  status?: MemoryStatus;
  limit?: number;
} = {}): Promise<MemoryList> {
  const { data } = await api.get<MemoryList>('/api/v1/memories', {
    params: {
      scope_type: params.scopeType,
      scope_id: params.scopeId,
      kind: params.kind,
      status: params.status,
      limit: params.limit,
    },
  });
  return data;
}

export async function updateMemory(memoryId: string, payload: UpdateMemoryRequest): Promise<Memory> {
  const { data } = await api.patch<Memory>(`/api/v1/memories/${memoryId}`, payload);
  return data;
}

export async function forgetMemory(memoryId: string): Promise<Memory> {
  const { data } = await api.delete<Memory>(`/api/v1/memories/${memoryId}`);
  return data;
}

export async function listConversationMemoryMounts(
  conversationId: string,
  limit = 50,
): Promise<MemoryMountList> {
  const { data } = await api.get<MemoryMountList>(
    `/api/v1/conversations/${conversationId}/memory-mounts`,
    { params: { limit } },
  );
  return data;
}

export async function getConversationMemoryHub(
  conversationId: string,
): Promise<ConversationMemoryHub> {
  const { data } = await api.get<ConversationMemoryHub>(
    `/api/v1/conversations/${conversationId}/memory-hub`,
  );
  return data;
}
